sdk/src/slot_history.rs

```diff
@@ -31,11 +46,37 @@ pub enum Check {
 
 impl SlotHistory {
     pub fn add(&mut self, slot: Slot) {
-        for skipped in self.next_slot..slot {
-            self.bits.set(skipped % MAX_ENTRIES, false);
+        if slot > self.next_slot && slot - self.next_slot >= MAX_ENTRIES {
+            // Wrapped past current history,
+            // clear entire bitvec.
+            let full_blocks = (MAX_ENTRIES as usize) / 64;
+            for i in 0..full_blocks {
+                self.bits.set_block(i, 0);
+            }
+            self.next_slot = slot + 1;
+        } else if slot >= self.next_slot {
+            // Current and next pointers are within
+            // the same MAX_ENTRIES slots
+            let start = self.next_slot % MAX_ENTRIES;
+            let end = slot % MAX_ENTRIES;
+            if end >= start {
+                // Straightforward clear between old head and new head
+                for skipped in start..end {
+                    self.bits.set(skipped, false);
+                }
+            } else {
+                // Wrapped case, need to clear the old head to the end
+                // and from 0 to the new head.
+                for skipped in 0..end {
+                    self.bits.set(skipped, false);
+                }
+                for skipped in start..MAX_ENTRIES {
+                    self.bits.set(skipped, false);
+                }
+            }
+            self.next_slot = slot + 1;
```

Comment:
I think it's better off to add an `else` clause with a panic like this:

```rust
} else if slot >= self.next_slot {
    ...
} else {
    assert!("history going backwards from {} to {}", self.next_slot, slot);
}
```

Reply:
Or can go backwards actually? it seems that previous code could for that, maybe unintentionally?

Reply:
it seems now we have tests: https://github.com/solana-labs/solana/pull/10175/files#diff-b6bc75c0bdcd2062ec00c3da3d974053R195 :)

Reply:
Or should we preserve the old behavior?: https://github.com/solana-labs/solana/pull/10175/files#r429238052

Reply:
Yea, I thought about that. I think the slot is always increasing, so the behavior actually doesn't matter. As you noticed the old implementation didn't make sense if you set an older slot.  It resets the next_slot and treats previously set slots in the future.

I wanted to have it make sense, but I'm open to preserving the previous behavior if we think it will break anything.